File: ble_coms/json_serializer.py
# json_serializer.py: establishes client/broker connection and parses/constructs payloads
import json # allow payloads to be written in strings
import paho.mqtt.client as mqtt # establish the python script as a client of the broker
import config
import bluetti_services
import time

# establish a client to pair to broker:
mqtt_client = mqtt.Client(
    mqtt.CallbackAPIVersion.VERSION2
)
# enter credentials manually for now:
mqtt_client.username_pw_set(
    config.MQTT_USER,
    config.MQTT_PASSWORD
)
# connect the new client to the broker:
mqtt_client.connect(
    config.MQTT_HOST,
    config.MQTT_PORT
)

mqtt_client.loop_start() # start the background network loop 

# stores the result of the most recent action performed on the battery
# so the controller knows if the last command succeeded/failed
class ActuatorState:
    def __init__(self):
        # default values until a command changes them
        self.request = "read" # overwrite with write/read
        self.target = None # overwrite with command/telemetry target
        self.value = None # no value being written/read
        self.action = None # overwrite with true/false
        self.flag = "idle" # overwrite with success/fail

actuator = ActuatorState()

# publish on one MQTT stream for the telemetry (bluetti/ac200/telemetry):
def json_publish_battery_state(state_payload):
    print("Publishing telemetry...")
    payload = {
        "device": "bluetti_ac200",
        "timestamp": time.time(),
        "action": {
            "request": actuator.request,
            "target": actuator.target,
            "value": actuator.value,
            "status": actuator.action,
            "state": actuator.flag,
        },
        "telemetry": {
            "home_telemetry": {
                "soc": state_payload["soc"],
                "voltage": state_payload["voltage"],
                "current": state_payload["current"],
                "pv_watts": state_payload["pv_watts"],
                "grid_watts": state_payload["grid_watts"],
                "ac_watts": state_payload["ac_watts"],
                "dc_watts": state_payload["dc_watts"],
            },
            "dc_output": {
                "value": state_payload["dc_output"],
            },
            "ac_output": {
                "value": state_payload["ac_output"],
            },
            "dc_input": {
                "value": state_payload["dc_input"],
            },
            "ac_input": {
                "value": state_payload["ac_input"],
            },
            "grid_permission": {
                "value": state_payload["grid_permission"],
            },
            "ac_input_setting": {
                "value": state_payload["ac_input_setting"],
            },
        },
    }
    print(json.dumps(payload, indent=2))
    # convert the dictionary into JSON and publish it to the broker
    return mqtt_client.publish(
    config.TOPIC_TELEMETRY,
    json.dumps(payload)
    )
# serialize schedule data into JSON, placeholder until scheduling is finished
def json_publish_battery_schedule(request, value, action,state, flag): #in state: 0 for reading, 1 for writing
    payload = {
        "device":"bluetti_ac200",
        "timestamp": time.time(),
        "action": {
                    "request": request, # what the controller wants to be written/read
                    "target": actuator.target, # the target of the command
                    "value": value, # what is being written/read (for validation)
                    "status": action, # action defined as read write or none
                    "state": flag # success or failure of ble coms
                },
        "schedule":{
            "grid_charging_permission":{
                "value": value.grid_charging_permission,
                "action": state.grid_charging_permission,
            },
            "ac_input_setting": {
                    "value": value.ac_input_setting,
                    "action": state.ac_input_setting,
        }
    }
}
    
    # serialize the dictionary into JSON and publish it to the telemetry topic
    return mqtt_client.publish(
        config.TOPIC_TELEMETRY,
        json.dumps(payload)
    )

# call in loop.py with:
#parsed = mqtt_service.latest_parsed
#if parsed is not None:
    #mqtt_service.update_battery_state(
        #parsed,
        #value,
        #state,)
# shared state updated by MQTT callbacks loop.py can read these
latest_parsed = None
latest_command = None
latest_schedule = None
# callback executed whenever a subscribed topic receives a new MQTT message
def on_message(client, userdata, msg):
    global latest_parsed
    global latest_command
    global latest_schedule

    if msg.topic == config.TOPIC_TELEMETRY:
        latest_parsed = json_process_telemetry(msg)

    elif msg.topic == config.TOPIC_COMMAND:
        latest_command = json_process_command(msg)

    elif msg.topic == config.TOPIC_SCHEDULE:
        latest_schedule = json_process_schedule(msg)
        # later:
        # send this command to loop.py

#call : parsed = json_process_telemetry(msg)

# deserialize an incoming telemetry packet back into a Python dictionary that is easier to work with
def json_process_telemetry(msg):
 # action defined as read write or none
    if not msg.payload:
        return None # ignore empty packets
    payload = json.loads(msg.payload)
    # safely retrieve nested dictionaries, .get() avoids exceptions if a field is missing
    telemetry = payload.get("telemetry",{})
    action = payload.get("action",{})
    home = telemetry.get("home_telemetry",{})
    # flatten the nested JSON into something easier to access throughout the rest of the program
    return {
        "action": {
            "request": action.get("request"),
            "target": action.get("target"),
            "value": action.get("value"),
            "status": action.get("status"),
            "state": action.get("state")
        },
        "soc": home.get("soc"),
        "voltage": home.get("voltage"),
        "current": home.get("current"),
        "pv_watts": home.get("pv_watts"),
        "grid_watts": home.get("grid_watts"),
        "ac_watts": home.get("ac_watts"),
        "dc_watts": home.get("dc_watts"),

        "AC output": telemetry.get("ac_output"),
        "AC input": telemetry.get("ac_input"),
        "AC input setting": telemetry.get("ac_input_setting"),
        "DC output": telemetry.get("dc_output"),
        "DC input": telemetry.get("dc_input"),
    }

# deserialize an incoming controller command
def json_process_command(msg):
    if not msg.payload:
        return None
    payload = json.loads(msg.payload)
    return {
        "request": payload.get("request"),
        "target": payload.get("target"),
        "value": payload.get("value"),
    }

def json_process_schedule(msg):
    if not msg.payload:
            return None
    payload = json.loads(msg.payload)
    return {
    "request": payload.get("request"),
    "charge_start": payload.get("charge_start"),
    "charge_end": payload.get("charge_end"),
    "discharge_start": payload.get("discharge_start"),
    "discharge_end": payload.get("discharge_end"),
}

#print(f"[READ] {label}: {values}", flush=True)

#for inside main py loop:


#if command is not None:
# if command["request"] == "read":
# continue polling
#if action == "write":
# perform ble write

# copy values from the parsed JSON into the battery objects so the rest of the program always uses newest data
def update_battery_state(parsed, value):

    value.soc = parsed["soc"]
    value.voltage = parsed["voltage"]
    value.current = parsed["current"]
    value.pv_watts = parsed["pv_watts"]
    value.grid_watts = parsed["grid_watts"]
    value.ac_watts = parsed["ac_watts"]
    value.dc_watts = parsed["dc_watts"]

    value.ac_output = parsed["AC output"]["value"]
    value.ac_input = parsed["AC input"]["value"]
    value.ac_input_setting = parsed["AC input setting"]["value"]
    value.dc_output = parsed["DC output"]["value"]
    value.dc_input = parsed["DC input"]["value"]

# register the callback and listen for both telemetry and controller command topics
mqtt_client.on_message = on_message

mqtt_client.subscribe(config.TOPIC_TELEMETRY)
mqtt_client.subscribe(config.TOPIC_COMMAND)
mqtt_client.subscribe(config.TOPIC_SCHEDULE)

File: src/json_serializer.cpp

#include "json_serializer.h"
#include <iostream>
#include <chrono>
#include <nlohmann/json.hpp> //JSON parsing library
using json = nlohmann::json;

bool ParseBatteryTelemetry(const std::string& payload,DeviceInformation& battery){
    try{
        //convert the MQTT payload string into a JSON object
        json data=json::parse(payload);
        //copy telemetry values from the JSON packet into the battery structure
        battery.timestamp = data["timestamp"];
        battery.device = data["device"];

        // handle null values during JSON parse
        auto action=data.value("action",json::object());
        battery.last_request=action.value("request","");
        if(action["target"].is_string())
            battery.last_target=action["target"].get<std::string>();
        else
            battery.last_target="";
        if(action["value"].is_number())
            battery.last_value=action["value"].get<int>();
        else
            battery.last_value=0;
        if(action["status"].is_boolean())
            battery.last_status=action["status"].get<bool>();
        else
            battery.last_status=false;
        battery.last_state=action.value("state","");

        battery.soc= data["telemetry"]["home_telemetry"]["soc"];
        battery.voltage = data["telemetry"]["home_telemetry"]["voltage"];
        battery.current = data["telemetry"]["home_telemetry"]["current"];
        battery.pv_watts = data["telemetry"]["home_telemetry"]["pv_watts"];       
        battery.grid_watts = data["telemetry"]["home_telemetry"]["grid_watts"];
        battery.ac_watts = data["telemetry"]["home_telemetry"]["ac_watts"];
        battery.dc_watts = data["telemetry"]["home_telemetry"]["dc_watts"];
        
        battery.grid_permission = data["telemetry"]["grid_permission"]["value"];

        battery.ac_output= data["telemetry"]["ac_output"]["value"];
        battery.dc_output= data["telemetry"]["dc_output"]["value"];
        battery.ac_input= data["telemetry"]["ac_input"]["value"];
        battery.dc_input= data["telemetry"]["dc_input"]["value"];
        battery.ac_input_setting= data["telemetry"]["ac_input_setting"]["value"];
        return true;
    }
    catch(const json::exception& e){
        std::cerr<<"JSON parse failed: "<<e.what()<<std::endl;
        return false;
    }
}

std::string BuildBatteryCommandJson(const BatteryCommand& command){
    json payload; // create an empty JSON object

    payload["device"] = "bluetti_ac200"; // device information
    payload["timestamp"] = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch()).count(); // time the command was created

    //controller commands
    payload["request"] = command.request;
    payload["target"] = command.target;
    payload["value"] = command.value;

    //convert the JSON object into a string
    return payload.dump();
}

std::string BuildBatteryScheduleJson(
    const BatterySchedule& schedule){
    json payload;

    payload["device"] = "bluetti_ac200";
    payload["timestamp"] = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch()).count();
    payload["request"] = schedule.request;

    payload["charge_start"] = schedule.charge_start;
    payload["charge_end"] = schedule.charge_end;

    payload["discharge_start"] = schedule.discharge_start;
    payload["discharge_end"] = schedule.discharge_end;

    return payload.dump();
}